### PR TITLE
Feature/machine controller

### PR DESCRIPTION
**Pull request by kyle-co**
Machine controller implementation, and other changes. They are largely alignment & cleanup efforts, no functional changes besides the new machine controller.

### CODE REVIEW

File: machinecontroller.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+		return -1, err
+	}
+
+	nextID, err := strconv.ParseInt(NextID.Data.NextID[0], 10, 32)
+	if err != nil {
+		return -1, err
+	}
+
+	// ParseInt returns int64 -- cast to int32 to maintain consistency with models in API docs
+	return int32(nextID), err
```
Comment (zachflower) [inline machinecontroller.go:252]: This general pattern seems to be repeated pretty frequently. Is it possible to DRY it out, or do you prefer to lean into the explicitness?

Comment (kyle-co) [reply machinecontroller.go]: Thanks for checking this out, glad to have your input. 

I'll do a union of the patterns and see what common denominators can be extracted. We would probably need a pattern for each HTTP method. A single func to handle all HTTP funcs would probably be needlessly complex.

File: liqid.go [hunk 2]
```diff
@@ -25,8 +25,8 @@ func (client *LiqidClient) Init() error {
 			return errors.New("LIQID_HOST is not set")
 		}
 	}
-	client.BaseURL = fmt.Sprintf("%s/%s", client.Host, APIPATH)
-	client.HTTPClient = &http.Client{Timeout: time.Second * 20}
+	client.BaseURL = fmt.Sprintf("%s/%s", client.Host, APIURL)
+	client.HTTPClient = &http.Client{Timeout: time.Second * 5}
```
Comment (schittumuru) [inline liqid.go:29]: Can we externalise the timeout value either passing it as as parameter to the Init() method or environment variable ?

File: machinecontroller.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
+package liqid_go
```
Comment (schittumuru) [inline machinecontroller.go:1]: As per package naming conventions its good to have a single word package names, without any special characters.

File: machinecontroller.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+	"fmt"
+	"net/http"
+	"strconv"
+)
+
+const (
+	MachineURL        = "machine"
+	MachineDetailsURL = "machine/details"
+	NextMachineURL    = "machine/nextid"
+	MachineP2PURL     = "machine/p2p"
```
Comment (schittumuru) [inline machinecontroller.go:15]: Can we have these constant variables names in UPPERCASE, so that it would be to distinguish from normal variables?

File: sshcontroller.go [hunk 5]
```diff
@@ -7,24 +7,24 @@ import (
 	"net/http"
 )
 
-const SSHENABLEDPATH = "/ssh/enable"
+const SSHEnabledURL = "ssh/enable"
```
Comment (schittumuru) [inline sshcontroller.go:10]: Same as above, can we use UPPERCASE to declare constants ?

File: statecontroller.go [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
-	DEGRADED      = "/degraded"
-	CLEARDEGRADED = "/degraded/clear"
-	FLAGS         = "/flags"
+	StateURL         = "state"
+	FabricURL        = "fabric"
+	GroupURL         = "fabric/group"
+	MachineStateURL  = "fabric/group/machine"
+	DegradedURL      = "degraded"
+	ClearDegradedURL = "degraded/clear"
+	FlagsURL         = "flags"
```
Comment (schittumuru) [inline statecontroller.go:10]: Same as above, can we use UPPERCASE to declare constants ?

File: systemcontroller.go [hunk 7]
```diff
@@ -6,12 +6,12 @@ import (
...
-	RESTARTSWITCH    = "system/restart/switch"
-	SHUTDOWN         = "system/shutdown"
-	SHUTDOWNNODE     = "system/shutdown/%d/%d/%d" // "/shutdown/{rack}/{shelf}/{node}
-	STATERESET       = "system/state/reset"
+	RestartFabricURL    = "system/restart/fabric"
+	RestartHierarchyURL = "system/restart/hierarchy"
+	RestartSwitchURL    = "system/restart/switch"
+	ShutdownURL         = "system/shutdown"
+	ShutdownNodeURL     = "system/shutdown/%d/%d/%d" // "/shutdown/{rack}/{shelf}/{node}
+	StateResetURL       = "system/state/reset"
```
Comment (schittumuru) [inline systemcontroller.go:14]: Same as above, can we use UPPERCASE to declare constants ?

File: sshcontroller.go (hunk 5 above)
Comment (kyle-co) [reply sshcontroller.go:10]: Uppercase constants are against golang standards.

Comment (kyle-co) [reply sshcontroller.go:10]: I'm willing to change them if there's enough of a style preference, but there are a number of discussions dissuading the use of uppercase.

Comment (kyle-co) [reply sshcontroller.go:10]: https://stackoverflow.com/questions/22688906/go-naming-conventions-for-const

File: liqid.go (hunk 2 above)
Comment (kyle-co) [reply liqid.go:29]: We can do that, sure.

File: machinecontroller.go (hunk 3 above)
Comment (kyle-co) [reply machinecontroller.go]: Fixed.

File: machinecontroller.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+		return -1, err
+	}
+	nextID, err := strconv.ParseInt(jsonResponse.Data.NextID[0], 10, 32)
+	return int32(nextID), err
+}
+
+// EnableP2P Enable P2P on specified machine
+func (client *LiqidClient) EnableP2P(mach Machine) error {
+	return client.DoAll(http.MethodPost, nil, mach, client.BaseURL, MachineP2PURL)
+}
```
Comment (schittumuru) [inline machinecontroller.go:151]: After going through the couple of controller classes, I got a feeling that all the methods in all controllers are implemented by LiqidClient and are accessible by the user of SDK  by using one single instance of LiqidClient type. 
I am thinking if we can have separate types for each controller and all the methods of that controller are accessible by using that type. 

We can use LiqidClient as a wrapper class which makes actual http calls.

Comment (kyle-co) [reply machinecontroller.go:151]: I appreciate you getting your hands into the repo and your willingness to contribute ideas. Can we talk about them and capture them in some separate tickets?

File: liqid.go [hunk 9]
```diff
@@ -25,15 +29,103 @@ func (client *LiqidClient) Init() error {
...
+
+		resp, err := client.HTTPClient.Do(req)
+		if err != nil {
+			return err
+		}
+		return resp.Body.Close()
+
+	default:
+		return errors.New("unsupported HTTP method")
+	}
```
Comment (schittumuru) [inline liqid.go:130]: Can we organise this client a little bit with separate methods like doGet, doPost ..etc instead of handling the logic with a switch case ? 
May be in separate pull request ? Please share your thoughts.

Comment (kyle-co) [reply liqid.go:130]: Yea, I'm open to that.

Comment (kyle-co) [reply liqid.go:130]: Let's wrangle that in another PR.

Comment (zachflower) [reply liqid.go:130]: Agreed, we can tackle it later, but I do like the idea of having those as helper methods (the original `do` with additional `doGet`, `doPut`, etc). It's a pretty common pattern in other HTTP client libraries I've seen.

**Review by zachflower — APPROVED** [review]
This is awesome work, the new `do` method _really_ DRYs things up a lot. Well done bud.
